### PR TITLE
[MRG+1] Infer number of records from file size when n_records == -1

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/4028.

I'll add the test file tomorrow.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Actually, we could simulate this in testing with something like:
```
with open(edf_fname, 'rb') as fid_in:
    with open(broken_fname, 'wb') as fid_out:
        fid_in.seek(0, 2)
        n_bytes = fid.tell()
        fid_in.seek(0, 0)
        fid_out.write(fid_in.read(int(n_bytes * 0.8)))
```
Then you can test to see if `broken_fname` can be read properly.

**Comment by jaeilepp** [comment]
The problem with that is that the header remains the same. I would not do this check unless it is necessary (n_records == -1). I tried your code and the ``broken_fname`` was read correctly when I removed the conditional.

I now realize that the heading of this PR is a bit wrong since the header is not corrupted. It's just that the ``n_records`` is not set to the number of records.

**Comment by jaeilepp** [comment]
@jpazdera and @cbrnr can you check if this fix works for your files? The only effective difference is the usage of sum of ``n_samps`` to define the length of the record. It should be more robust for variable sampling rates.

@pazdera is it okay if add the ``numrecs-1.bdf`` file you provided to the testing data?

**Comment by cbrnr** [comment]
@jaeilepp your fix works for my file! Nice! Thanks for taking care of this! 😃

**Comment by jaeilepp** [comment]
I added the test utilizing a simulated test file.

File: mne/io/edf/edf.py [hunk 1]
```diff
@@ -380,6 +380,17 @@ def _get_edf_info(fname, stim_channel, annot, annotmap, eog, misc, exclude,
...
+        if n_records == -1:
+            warn('Unable to read the number of records from the header '
+                 '(perhaps the recording was not stopped before exiting). '
+                 'Inferring from the file size.')
+            fid.seek(0, 2)
+            n_bytes = fid.tell()
+            n_data_bytes = n_bytes - header_nbytes
+            total_samps = (n_data_bytes // 3 if subtype == '24BIT'
+                           else n_data_bytes // 2)
+            edf_info['n_records'] = n_records = total_samps // np.sum(n_samps)
```
Comment (Eric89GXL) [inline mne/io/edf/edf.py]: It probably makes sense to just always do this calculation, and `warn` if `n_records` is missing or there is a mismatch between expected and calculated

**Comment by Eric89GXL** [comment]
LGTM +1 for merge

**Comment by agramfort** [comment]
thx @jaeilepp !

**Pull request merged**